spec/syntax.md

```diff
@@ -229,272 +314,145 @@ A message can have multiple selectors.
+> {This placeholder references a {$variable}}
+> {This placeholder references a function on a variable: {$variable :function with=options}}
 > ```
 
-### Variants
+### Operand
 
-A **_variant_** is a _pattern_ associated with a set of _keys_.
-Each _variant_ MUST begin with the _keyword_ `when`,
-be followed by a non-empty sequence of _keys_,
-and terminate with a valid _pattern_.
-The key `*` is a "catch-all" key, matching all selector values.
+An **_operand_** is a _literal_ or a _variable_ to be evaluated in an _expression_.
+An _operand_ MAY optionally be followed by an _annotation_.
 
 ```abnf
-variant = when 1*(s key) [s] pattern
-key     = literal / "*"
+operand    = literal / variable
 ```
 
-A _well-formed_ message is considered _valid_ if the following requirements are satisfied:
-
-- The number of keys on each _variant_ MUST be equal to the number of _selectors_.
-- At least one _variant's_ keys MUST all be equal to the catch-all key (`*`).
-- Each _selector_ MUST have an _annotation_,
-  or contain a _variable_ that directly or indirectly references a _declaration_ with an _annotation_.
-
-### Patterns
+### Annotation
 
-A **_pattern_** is a sequence of translatable elements.
-Patterns MUST be delimited with `{` at the start, and `}` at the end.
-A _pattern_'s contents MAY be empty.
-Whitespace within a _pattern_ is meaningful and MUST be preserved.
-This serves 3 purposes:
-
-- The message can be unambiguously embeddable in various container formats
-  regardless of the container's whitespace trimming rules.
-  E.g. in Java `.properties` files,
-  `hello = {Hello}` will unambiguously define the `Hello` message without the space in front of it.
-- The message can be conveniently embeddable in various programming languages
-  without the need to escape characters commonly related to strings, e.g. `"` and `'`.
-  Such need might still occur when a single or double quote is
-  used in the translatable content.
-- The syntax needs to make it as clear as possible which parts of the message body
-  are translatable and which ones are part of the formatting logic definition.
+An **_annotation_** is part of an _expression_ containing either 
+a _function_ together with its associated _options_, or
+a _private-use_ or _reserved_ sequence.
 
 ```abnf
-pattern = "{" *(text / expression) "}"
+annotation = (function *(s option)) / reserved / private-use
 ```
 
-> **Example**
-> 
-> A simple _pattern_ containing _text_:
-> ```
-> {Hello, world!}
-> ```
->
-> An empty _pattern_:
-> ```
-> {}
-> ```
->
-> Some _patterns_ with _expressions_:
-> ```
-> {{$foo}}
-> {Hello {$user}!}
-> {You sent {$count :number maxFractionDigits=0} notifications to {$numFriends :number type=spellout} friends.}
-> ```
->
-> A _pattern_ containing three spaces:
-> ```
-> {   }
-> ```
+### Function
 
-### Expressions
+A **_function_** is a used to evaluate, format, select, or otherwise process an _operand_,
+or, if lacking an _operand_, its _options_.
```

Comment:
This implies that a function with an operand does not care about its options, and that a function without an operand requires options. The previous reference to processing its annotation was really clumsy, but less incorrect.

This sentence should probably be completely refactored.

Reply:
Note that this sentence doesn't make sense as written. Refactored this section.